### ISSUE TITLE
Redefinition of TypeScript's builtin `Partial<T>`

### ISSUE THREAD

**Issue by T99** [elijahjcobb/silicon#2]
https://github.com/elijahjcobb/silicon/blob/b64b73d8fd0e18ba4dc05d808035a62d25b5ca2d/ts/object/SiObject.ts#L15

```typescript
// TypeScript's builtin type:
type Partial<T> = { [P in keyof T]: T[P] | undefined };
```

**Comment by elijahjcobb** [comment]
Although at first glance this looks like a redefinition, it is not. TypeScript's builtin `Partial<T>` is defined as `type Partial<T> = { [P in keyof T]?: T[P]; };` where as `SiPartial` is defined as `type SiPartial<T extends object> = { [P in keyof T]: T[P] | undefined; }`.

The key difference is in TypeScript's partial defined that every key can exist or not exist. However, in the `SiPartial` every key **MUST** exist and the underlying value can either be defined or not. The reason I had to do this is so that calling something like: `X.props.name` would return `string | undefined` and not `string`.

**Comment by T99** [comment]
I'm almost certain that...
```typescript
value?: Type
```
...is _exactly_ the same as...
```typescript
value: Type | undefined
```
...with strict-null-checks.

See [this](https://stackoverflow.com/questions/47942141/optional-property-class-in-typescript) Stack post.

**Comment by elijahjcobb** [comment]
Yes you are right. Fixed in commit `753bbab`.
